shared/src/balancer_event_handler.rs

```diff
@@ -3,195 +3,183 @@ use crate::{
-        {
-            return Some(WeightedPool {
-                pool_id: pool_registration.pool_id,
-                pool_address: pool_registration.pool_address,
-                tokens: tokens_registration.tokens,
-                specialization: pool_registration.specialization,
-                block_created: self.block_created,
-            });
-        }
-        None
+    async fn into_pool(self, data_fetcher: &dyn PoolDataFetching) -> Result<WeightedPool> {
+        let pool_address = self.pool_creation.pool_address;
+        let pool_data = data_fetcher.get_pool_data(pool_address).await?;
+        return Ok(WeightedPool {
+            pool_id: pool_data.pool_id,
+            pool_address,
+            tokens: pool_data.tokens,
+            normalized_weights: pool_data.weights,
+            block_created: self.block_created,
+        });
+    }
+}
+
+#[derive(Clone)]
+pub struct WeightedPoolData {
+    pool_id: H256,
+    tokens: Vec<H160>,
+    weights: Vec<U256>,
+}
+
+#[automock]
+#[async_trait::async_trait]
+trait PoolDataFetching: Send + Sync {
+    async fn get_normalized_weights(&self, pool_address: H160) -> Result<Vec<U256>>;
+    async fn get_pool_tokens(&self, pool_id: H256) -> Result<Vec<H160>>;
+    async fn get_pool_id(&self, pool_address: H160) -> Result<H256>;
+    async fn get_pool_data(&self, pool_address: H160) -> Result<WeightedPoolData>;
+}
+
+#[async_trait::async_trait]
+impl PoolDataFetching for Web3 {
+    async fn get_normalized_weights(&self, pool_address: H160) -> Result<Vec<U256>> {
+        let pool_contract = BalancerV2WeightedPool::at(self, pool_address);
+        Ok(pool_contract
+            .methods()
+            .get_normalized_weights()
+            .call()
+            .await?)
+    }
+
+    async fn get_pool_tokens(&self, pool_id: H256) -> Result<Vec<H160>> {
+        let vault = BalancerV2Vault::deployed(&self).await?;
+        // Note, this is also recovering the balances.
+        Ok(vault
+            .methods()
+            .get_pool_tokens(Bytes(pool_id.0))
+            .call()
+            .await?
+            .0)
+    }
+
+    async fn get_pool_id(&self, pool_address: H160) -> Result<H256> {
+        let pool_contract = BalancerV2WeightedPool::at(self, pool_address);
+        Ok(H256::from(
+            pool_contract.methods().get_pool_id().call().await?.0,
+        ))
+    }
+
+    async fn get_pool_data(&self, pool_address: H160) -> Result<WeightedPoolData> {
+        let pool_id = self.get_pool_id(pool_address).await?;
+        let tokens = self.get_pool_tokens(pool_id).await?;
+        let weights = self.get_normalized_weights(pool_address).await?;
+        Ok(WeightedPoolData {
+            pool_id,
+            tokens,
+            weights,
+        })
     }
 }
 
 /// The BalancerPool struct represents in-memory storage of all deployed Balancer Pools
-#[derive(Debug, Default)]
+#[derive(Derivative)]
+#[derivative(Debug)]
 pub struct BalancerPools {
```

Comment:
Maybe this struct should now become a WeightedBalancerPoolRegistry?

Reply:
Yes, I was trying to keep this massive PR as small as possible and would have done this in a separate PR.

Reply:
I actually like `BalancerPools`, as we can extend this to instead of return a:
```
enum BalancerPool {
    Weighted(...),
    Stable(...),
}
```

When we start supporting more pool types.

Reply:
Cool idea.